File: frontend/src/types/route.ts
// Route related types

import { Coordinates, OptimizationCriteria } from './common';

export interface RouteRequest {
  origin: string | Coordinates;
  destination: string | Coordinates;
  vehicle_id: string;
  alternatives?: number;
  avoid_tolls?: boolean;
  optimize_for?: OptimizationCriteria;
}

export interface RouteResponse {
  request_id: string;
  recommended_route: Route;
  alternative_routes: Route[];
  savings_analysis: SavingsAnalysis;
  calculation_time_ms: number;
  timestamp: string;
}

export interface Route {
  id: string;
  geometry: GeoJSONLineString;
  distance: number; // km
  duration: number; // minutes
  cost_breakdown: CostBreakdown;
  toll_points: TollPoint[];
  route_type: RouteType;
}

export interface CostBreakdown {
  fuel_cost: number;
  toll_cost: number;
  total_cost: number;
  fuel_liters: number;
  toll_count: number;
}

export interface TollPoint {
  toll_id: string;
  name: string;
  coordinates: Coordinates;
  tariff: number;
  route_code: string;
}

export interface SavingsAnalysis {
  recommended_route_id: string;
  fastest_route_cost: number | null;
  cheapest_route_cost: number;
  savings_amount: number | null;
  savings_percentage: number | null;
  comparison_summary: string;
}

export type RouteType = 'recommended' | 'fastest' | 'cheapest' | 'alternative' | 'scenic';

export interface GeoJSONLineString {
  type: 'LineString';
  coordinates: [number, number][]; // [longitude, latitude]
}

export interface RouteHistory {
  id: string;
  user_id: string;
  vehicle_id: string;
  company_id: string;
  origin_address: string;
  destination_address: string;
  origin_coordinates: Coordinates;
  destination_coordinates: Coordinates;
  selected_route: Route;
  total_distance: number;
  total_duration: number;
  fuel_cost: number;
  toll_cost: number;
  total_cost: number;
  savings_amount: number;
  created_at: string;
}

export interface RouteHistoryResponse {
  routes: RouteHistory[];
  total: number;
  page: number;
  size: number;
  pages: number;
}

export interface RouteStatistics {
  total_routes: number;
  total_distance: number;
  total_cost: number;
  total_savings: number;
  average_distance: number;
  average_cost: number;
  most_used_vehicle: string | null;
  most_common_origin: string | null;
  most_common_destination: string | null;
  date_range: {
    start: string;
    end: string;
  };
}

export interface GeocodingResponse {
  results: GeocodingResult[];
  status: 'success' | 'no_results' | 'error';
}

export interface GeocodingResult {
  address: string;
  coordinates: Coordinates;
  formatted_address: string;
  confidence: number;
  place_type: string;
}

export interface RouteReportRequest {
  route_id: string;
  report_type: 'complete' | 'simple';
  include_map: boolean;
  format: 'pdf' | 'html';
}

export interface RouteReportResponse {
  report_id: string;
  download_url: string;
  expires_at: string;
  file_size: number;
  format: 'pdf' | 'html';
}

// Map related types
export interface MapViewState {
  longitude: number;
  latitude: number;
  zoom: number;
  bearing?: number;
  pitch?: number;
}

export interface MapBounds {
  north: number;
  south: number;
  east: number;
  west: number;
}

export interface RouteVisualizationOptions {
  showTolls: boolean;
  showAlternatives: boolean;
  highlightRecommended: boolean;
  showCostLabels: boolean;
}

// Route planning types
export interface RoutePlanningState {
  origin: {
    address: string;
    coordinates: Coordinates | null;
  };
  destination: {
    address: string;
    coordinates: Coordinates | null;
  };
  selectedVehicle: string | null;
  optimizationCriteria: OptimizationCriteria;
  avoidTolls: boolean;
  maxAlternatives: number;
}

export interface RouteCalculationProgress {
  step: 'geocoding' | 'routing' | 'cost_analysis' | 'complete';
  progress: number; // 0-100
  message: string;
}

File: backend/app/schemas/route.py
"""
Pydantic schemas for route calculation and response.
"""
from pydantic import BaseModel, Field, validator
from typing import Optional, List, Dict, Any, Union, Tuple
from datetime import datetime
from uuid import UUID
from decimal import Decimal


class Coordinates(BaseModel):
    """Geographic coordinates."""
    longitude: float = Field(..., ge=-180, le=180)
    latitude: float = Field(..., ge=-90, le=90)


class RouteRequest(BaseModel):
    """Schema for route calculation request."""
    origin: Union[str, Coordinates] = Field(..., description="Origin address or coordinates")
    destination: Union[str, Coordinates] = Field(..., description="Destination address or coordinates")
    vehicle_id: UUID = Field(..., description="Vehicle ID for cost calculation")
    alternatives: int = Field(default=3, ge=1, le=5, description="Number of route alternatives")
    avoid_tolls: bool = Field(default=False, description="Try to avoid toll roads")
    optimize_for: str = Field(default="cost", description="Optimization criteria: cost, time, distance")

    @validator('optimize_for')
    def validate_optimize_for(cls, v):
        """Validate optimization criteria."""
        allowed = ['cost', 'time', 'distance']
        if v not in allowed:
            raise ValueError(f'optimize_for must be one of: {allowed}')
        return v


class TollPoint(BaseModel):
    """Toll point information in route."""
    toll_id: UUID
    name: str
    coordinates: Coordinates
    tariff: Decimal
    route_code: Optional[str]


class CostBreakdown(BaseModel):
    """Cost breakdown for a route."""
    fuel_cost: Decimal = Field(..., description="Fuel cost in ARS")
    toll_cost: Decimal = Field(..., description="Toll cost in ARS")
    total_cost: Decimal = Field(..., description="Total cost in ARS")
    fuel_liters: float = Field(..., description="Fuel consumption in liters")
    toll_count: int = Field(..., description="Number of tolls on route")


class Route(BaseModel):
    """Individual route information."""
    id: str = Field(..., description="Route identifier")
    geometry: Dict[str, Any] = Field(..., description="GeoJSON LineString geometry")
    distance: float = Field(..., description="Distance in kilometers")
    duration: int = Field(..., description="Duration in minutes")
    cost_breakdown: CostBreakdown
    toll_points: List[TollPoint]
    route_type: str = Field(..., description="Route type: fastest, cheapest, balanced")


class SavingsAnalysis(BaseModel):
    """Savings analysis comparing routes."""
    recommended_route_id: str
    fastest_route_cost: Optional[Decimal]
    cheapest_route_cost: Decimal
    savings_amount: Optional[Decimal]
    savings_percentage: Optional[float]
    comparison_summary: str


class RouteResponse(BaseModel):
    """Schema for route calculation response."""
    request_id: str = Field(..., description="Unique request identifier")
    recommended_route: Route
    alternative_routes: List[Route]
    savings_analysis: SavingsAnalysis
    calculation_time_ms: int
    timestamp: datetime


class RouteCalculation(BaseModel):
    """Schema for storing route calculation results."""
    id: UUID
    user_id: UUID
    vehicle_id: UUID
    company_id: UUID
    origin_address: Optional[str]
    destination_address: Optional[str]
    origin_coordinates: Coordinates
    destination_coordinates: Coordinates
    selected_route: Route
    total_distance: float
    total_duration: int
    fuel_cost: Decimal
    toll_cost: Decimal
    total_cost: Decimal
    savings_amount: Optional[Decimal]
    created_at: datetime

    class Config:
        from_attributes = True


class RouteHistoryResponse(BaseModel):
    """Schema for route history responses."""
    routes: List[RouteCalculation]
    total: int
    page: int
    size: int
    pages: int


class RouteReportRequest(BaseModel):
    """Schema for route report generation request."""
    route_id: UUID
    report_type: str = Field(default="complete", description="Report type: complete, simple")
    include_map: bool = Field(default=True, description="Include map in report")
    format: str = Field(default="pdf", description="Report format: pdf, html")

    @validator('report_type')
    def validate_report_type(cls, v):
        """Validate report type."""
        allowed = ['complete', 'simple']
        if v not in allowed:
            raise ValueError(f'report_type must be one of: {allowed}')
        return v

    @validator('format')
    def validate_format(cls, v):
        """Validate report format."""
        allowed = ['pdf', 'html']
        if v not in allowed:
            raise ValueError(f'format must be one of: {allowed}')
        return v


class RouteReportResponse(BaseModel):
    """Schema for route report response."""
    report_id: str
    download_url: str
    expires_at: datetime
    file_size: int
    format: str


class RouteStatistics(BaseModel):
    """Schema for route statistics."""
    total_routes: int
    total_distance: float
    total_cost: Decimal
    total_savings: Decimal
    average_distance: float
    average_cost: Decimal
    most_used_vehicle: Optional[str]
    most_common_origin: Optional[str]
    most_common_destination: Optional[str]
    date_range: Dict[str, datetime]


class GeocodingResult(BaseModel):
    """Schema for geocoding results."""
    address: str
    coordinates: Coordinates
    formatted_address: str
    confidence: float
    place_type: str


class GeocodingResponse(BaseModel):
    """Schema for geocoding API response."""
    results: List[GeocodingResult]
    status: str